File: 김혜은/ac_high1.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class Combination {
    static void combination(List<Integer> arr, int r, int index, List<Integer> data, int i, int sum, int target) {
        if (index == r) {
            if (sum == target) {
                for (int j = 0; j < r; j++) {
                    System.out.print(data.get(j) + " ");
                }
                System.out.println();
            }
            return;
        }

        if (i >= arr.size()) {
            return;
        }

        data.set(index, arr.get(i));
        combination(arr, r, index + 1, data, i + 1, sum + arr.get(i), target);

        combination(arr, r, index, data, i + 1, sum, target);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter N: ");
        int N = scanner.nextInt();

        List<Integer> M = new ArrayList<>();
        System.out.print("Enter array elements: ");
        for (int i = 0; i < N; i++) {
            M.add(scanner.nextInt());
        }

        System.out.print("Enter target sum (L): ");
        int L = scanner.nextInt();

        List<Integer> data = new ArrayList<>(Arrays.asList(new Integer[N]));
        combination(M, N, 0, data, 0, 0, L);
    }
}

File: 김혜은/ac_high1.c
#include <stdio.h>
#include <stdlib.h>

void combination(int arr[], int n, int r, int index, int data[], int i, int sum, int target);

int main() {
    int N, L;

    printf("Enter the size of array: ");
    scanf("%d", &N);

    int *M = (int *)malloc(N * sizeof(int));

    printf("Enter array elements: ");
    for (int i = 0; i < N; i++) {
        scanf("%d", &M[i]);
    }

    printf("Enter target sum (L): ");
    scanf("%d", &L);

    int data[N];

    combination(M, N, N, 0, data, 0, 0, L);

    free(M);

    return 0;
}

void combination(int arr[], int n, int r, int index, int data[], int i, int sum, int target) {
    if (index == r) {
        if (sum == target) {
            for (int j = 0; j < r; j++) {
                printf("%d ", data[j]);
            }
            printf("\n");
        }
        return;
    }

    if (i >= n)
        return;

    data[index] = arr[i];
    combination(arr, n, r, index + 1, data, i + 1, sum + arr[i], target);

    combination(arr, n, r, index, data, i + 1, sum, target);
}
